File: src/main/java/com/ejo/game/scenes/GameScene.java
package com.ejo.game.scenes;

import com.ejo.game.render.Background;
import com.ejo.game.render.Ground;
import com.ejo.game.entity.Meteor;
import com.ejo.game.entity.MeteorSpawner;
import com.ejo.game.entity.Player;
import com.ejo.game.math.QuickTimer;
import com.ejo.game.math.Vector;

import javax.swing.*;
import java.awt.*;
import java.util.ArrayList;

public class GameScene extends Scene {

    private final QuickTimer countdownTimer;

    private int countdown = 3;

    private boolean started;
    private boolean gameOver;

    private final ArrayList<Meteor> meteorList;

    private final MeteorSpawner meteorSpawner;

    private final Background background;
    private final Ground ground;
    private final Player player;

    public GameScene() {
        super("Game Scene");
        this.countdownTimer = new QuickTimer();
        this.started = false;
        this.gameOver = false;

        this.meteorList = new ArrayList<>();

        this.background = new Background(100,10);
        this.ground = new Ground(500,30);
        this.player = new Player(new Vector(-100,-100));
        this.meteorSpawner = new MeteorSpawner(meteorList);
    }

    @Override
    public void draw(Graphics2D graphics) {
        background.draw(graphics,Vector.NULL);
        ground.draw(graphics,Vector.NULL);

        countdownStart(graphics);

        if (!started) return;
        if (gameOver) return;

        managePlayer(graphics);

        manageMeteors(graphics);
    }

    private void managePlayer(Graphics2D graphics) {
        if (!player.hasSpawned()) player.spawn();

        player.update();
        player.updateGroundCollision(ground);

        player.draw(graphics,player.getPos());
    }


    //TODO: Delete meteors when they hit the ground, Add an explosion animation too
    private void manageMeteors(Graphics2D graphics2D) {
        meteorSpawner.spawnMeteors(0,10,500);

        for (Meteor meteor : meteorList) {
            meteor.update();
            meteor.draw(graphics2D,meteor.getPos());
            if(player.isCollidingMeteor(meteor) && !gameOver) {
                gameOver = true;
                JButton retryButton = new JButton("Retry!");
                retryButton.setSize(200, 50);
                retryButton.setLocation((int)(getWindow().getSize().getX() / 2 - retryButton.getWidth() / 2), (int)(getWindow().getSize().getY() / 2 - retryButton.getHeight() / 2) + 10);
                retryButton.setFont(new Font("Arial",Font.PLAIN,30));
                retryButton.addActionListener(action -> getWindow().setScene(new GameScene()));
                addSwingComponents(retryButton);
            }
        }
    }


    private void countdownStart(Graphics2D graphics) {
        countdownTimer.start();
        if (!started) {
            graphics.setFont(new Font("Arial", Font.PLAIN, 50));
            graphics.drawString(String.valueOf(countdown), (int) getWindow().getSize().getX() / 2, (int) getWindow().getSize().getY() / 2);
        }
        if (countdownTimer.hasTimePassed(1000) && countdown >= 0) {
            if (countdown == 0) {
                started = true;
            } else {
                countdown--;
                countdownTimer.restart();
            }
        }
    }

}

File: src/scene/GameScene.py
from pygame import Color

from src.Window import Window
from src.entity.MeteorSpawner import MeteorSpawner, Meteor
from src.entity.Player import Player
from src.render.Button import Button
from src.render.Text import Text
from src.render.environment.Background import Background
from src.render.environment.Ground import Ground
from src.scene.Scene import Scene
from src.util.QuickTimer import QuickTimer
from src.util.Vector import Vector

class GameScene(Scene):

    def __init__(self, window : Window):
        super().__init__(window, "GameScene")

        #Game Starters
        self.is_started: bool = False
        self.countdown: int = 3
        self.countdown_timer: QuickTimer = QuickTimer()

        #Game Score
        self.score: int = 0
        self.score_timer: QuickTimer = QuickTimer()

        #Scene Drawables
        self.background: Background = Background(self,window,60,10)
        self.ground: Ground = Ground(self,75,30)

        #Entities
        self.meteor_spawner: MeteorSpawner = MeteorSpawner(self, Vector.null(), int(window.size.x), -20, int(window.size.y))
        self.player: Player = Player(self,Vector(window.size.x * .5 - 5,self.ground.pos.y - 30),10,Color(0,255,0))

        #Add All Drawables
        self.drawables.append(self.background)
        self.drawables.append(self.ground)
        self.drawables.append(self.player)

    def draw(self):
        self.update_countdown_start()

        if self.is_started:
            self.update_game()

        super().draw()

        self.draw_score_text()
        self.draw_countdown()

    # ======================

    # GAME UPDATE FUNCTIONS

    # ======================

    def update_game(self):
        self.meteor_spawner.update()
        self.player.update_player(self.ground.pos.y)

        #Update Meteor Collisions
        for meteor in self.drawables:
            if isinstance(meteor,Meteor):
                if meteor.is_colliding(self.player):
                    self.end_game()
                    self.drawables.remove(self.player)

        if self.is_started:
            self.score_timer.start()
            if self.score_timer.has_time_passed_s(1):
                self.score += 1
                self.score_timer.restart()


    def end_game(self):
        self.is_started = False
        action = lambda : self.window.set_scene(GameScene(self.window))
        retry_button: Button = Button(self,Vector.null(),Vector(250,100),Color(100,100,100),"Retry!",action)
        retry_button.pos = self.window.size * 0.5 - (retry_button.size * 0.5)
        self.drawables.append(retry_button)


    # ======================

    # COUNTDOWN & SCORE FUNCTIONS

    # ======================

    def draw_score_text(self):
        score_text: Text = Text(self,Vector(2,2),f"Score: {self.score}","Arial Black",20,Color(255,255,25))
        score_text.draw()

    def draw_countdown(self):
        if not self.is_started:
            text: Text = Text(self,Vector.null(),str(self.countdown),"Arial Black",50,Color(255,255,255))
            text.pos = (self.window.size * .5) - text.get_vector_size() * .5
            text.draw()

    def update_countdown_start(self):
        self.countdown_timer.start()
        if self.countdown_timer.has_time_passed_s(1) and self.countdown != "" and self.countdown >= 0:
            if self.countdown == 0:
                self.is_started = True
                self.countdown = "" #Converts the countdown to an empty string as to not draw it during src end
            else:
                self.countdown -= 1
            self.countdown_timer.restart()